### PR TITLE
ticdc: support detecting the Kafka cluster version automatically

### LINKED ISSUE #10852: TiCDC: kafka client has run out of available brokers to talk to: EOF
### What did you do?

create a changefeed 
`bin/cdc cli changefeed create -c test --sink-uri "kafka://127.0.0.1:9092/test?topic=test&protocol=open-protocol&max-message-bytes=41943040&compressionType=ZLIB"`

### What did you expect to see?

```
Create changefeed successfully!
ID: test
Info: {...}
```

### What did you see instead?

`Error: [CDC:ErrKafkaNewProducer]new kafka producer: kafka: client has run out of available brokers to talk to: EOF`

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
v7.6.0

```

Upstream TiKV version (execute `tikv-server --version`):

```console
v7.6.0
```

TiCDC version (execute `cdc version`):

```console
v7.6.0
```
Kafka version

```
v2.1.0
v2.1.1
v2.3.1
```

### PR DESCRIPTION
**Pull request by wk989898**
### What problem does this PR solve?

Issue Number: close #10852

### What is changed and how it works?
Add Kafka version detect. Connect a broker and use [`ApiVersionsRequest`](https://kafka.apache.org/protocol#api_versions) to get `ApiKeys` and `MaxVersion`, according to `MaxVersion` infer the Kafka version.

#### Test:
ticdc instance: v8.0.0-master-dirty
pd instance:v7.6.0
tikv instance:v7.6.0
tidb instance:v7.6.0
| kafka version |  status |
|   --- | ---     |
| 3.7.0 | success |
| 3.0.1 | success |
| 2.8.0 | success |
| 2.7.0 | success |
| 2.5.0 | success |
| 2.4.0 | success |
| 2.3.1 | success |
| 2.1.1 | success |
| 2.1.0 | success |
| 2.0.1 | success |
| 2.0.0 | success |
| 1.0.1 | success |
| 0.11.0 | success |
| 0.10.1 | [API is not supported](https://github.com/IBM/sarama/blob/f21c5125746f9d10fd731dfdff54a494098626d1/describe_configs_response.go#L136) |
| 0.10.0 | [API is not supported](https://github.com/IBM/sarama/blob/f21c5125746f9d10fd731dfdff54a494098626d1/describe_configs_response.go#L136) |
| 0.8.2 | not support [ApiVersionsRequest](https://kafka.apache.org/protocol#api_versions) |
#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

##### Will it cause performance regression or break compatibility?
not sure.
##### Do you need to update user documentation, design documentation or monitoring documentation?
no.
### Release note 

```release-note
`None`
```

### CODE REVIEW

File: pkg/sink/kafka/sarama.go [hunk 1]
```diff
@@ -122,6 +117,28 @@ func NewSaramaConfig(ctx context.Context, o *Options) (*sarama.Config, error) {
 		return nil, cerror.WrapError(cerror.ErrKafkaInvalidConfig, err)
 	}
 
+	kafkaVersion, err := GetKafkaVersion(ctx, config, o)
```
Comment (sdojjy) [inline pkg/sink/kafka/sarama.go:120]: move to line 131,   only detect the kafka cluster version when there is no version specified  by the user

File: pkg/sink/kafka/sarama.go [hunk 2]
```diff
@@ -122,6 +117,28 @@ func NewSaramaConfig(ctx context.Context, o *Options) (*sarama.Config, error) {
...
+			zap.String("defaultVersion", kafkaVersion.String()))
+	}
+
+	version, err := sarama.ParseKafkaVersion(o.Version)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidVersion, err)
+	}
+	if o.IsAssignedVersion {
+		config.Version = version
+		if version.String() != kafkaVersion.String() {
```
Comment (sdojjy) [inline pkg/sink/kafka/sarama.go:132]: according to the api version mapping, if the kafka cluster version is greater than 2.8.0,  `version.String() != kafkaVersion.String()` is always true.

File: pkg/sink/kafka/sarama.go [hunk 3]
```diff
@@ -167,3 +184,66 @@ func completeSaramaSASLConfig(ctx context.Context, config *sarama.Config, o *Opt
...
+		})
+	}
+	for i := range addrs {
+		broker := sarama.NewBroker(addrs[i])
+		err = broker.Open(config)
+		if err != nil {
+			log.Warn("Kafka fail to open broker", zap.String("addr", addrs[i]))
+			continue
+		}
+		defer func() {
```
Comment (sdojjy) [inline pkg/sink/kafka/sarama.go:206]: do not use defer in for loop

File: pkg/sink/kafka/sarama.go [hunk 4]
```diff
@@ -167,3 +184,66 @@ func completeSaramaSASLConfig(ctx context.Context, config *sarama.Config, o *Opt
...
+		// 0      Produce
+		// 3      Metadata (default)
+		version := apiResponse.ApiKeys[3].MaxVersion
+		kafkaVersion := requiredVersion(version)
+		return kafkaVersion, nil
+	}
+	return sarama.V2_0_0_0, err
+}
+
+func requiredVersion(version int16) sarama.KafkaVersion {
```
Comment (sdojjy) [inline pkg/sink/kafka/sarama.go:224]: `requiredVersion` is not a good name here, this function is  just a version mapping.

File: pkg/sink/kafka/sarama.go [hunk 5]
```diff
@@ -122,6 +117,28 @@ func NewSaramaConfig(ctx context.Context, o *Options) (*sarama.Config, error) {
...
+			zap.String("defaultVersion", kafkaVersion.String()))
+	}
+
+	version, err := sarama.ParseKafkaVersion(o.Version)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrKafkaInvalidVersion, err)
+	}
+	if o.IsAssignedVersion {
+		config.Version = version
+		if !version.IsAtLeast(sarama.V2_8_0_0) && version.String() != kafkaVersion.String() {
```
Comment (CharlesCheung96) [inline pkg/sink/kafka/sarama.go:132]: It is better to use constants rather than literals.

File: pkg/sink/kafka/sarama.go [hunk 6]
```diff
@@ -167,3 +184,68 @@ func completeSaramaSASLConfig(ctx context.Context, config *sarama.Config, o *Opt
...
+			addrs[i], addrs[j] = addrs[j], addrs[i]
+		})
+	}
+	for i := range addrs {
+		version, err := getKafkaVersionFromBroker(config, addrs[i])
+		if err == nil {
+			return version, err
+		}
+	}
+	return sarama.V2_0_0_0, err
```
Comment (CharlesCheung96) [inline pkg/sink/kafka/sarama.go:205]: It is better to use constants rather than literals.

File: pkg/sink/kafka/sarama.go [hunk 7]
```diff
@@ -167,3 +184,68 @@ func completeSaramaSASLConfig(ctx context.Context, config *sarama.Config, o *Opt
 
 	return nil
 }
+
+func getKafkaVersion(ctx context.Context, config *sarama.Config, o *Options) (sarama.KafkaVersion, error) {
+	var err error
+	addrs := o.BrokerEndpoints
+	if len(addrs) > 1 {
+		// Shuffle the list of addresses to randomize the order in which
```
Comment (3AceShowHand) [inline pkg/sink/kafka/sarama.go:192]: Why shuffle here ?

Comment (wk989898) [reply pkg/sink/kafka/sarama.go:192]: Prevent latency from increasing when the first address is inaccessible.

**Review by CharlesCheung96 — APPROVED** [review]

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**